llvm/lib/CheerpUtils/PartialExecuter.cpp

```diff
@@ -0,0 +1,1393 @@
+			case Instruction::InsertElement:
+			case Instruction::Store:
+			case Instruction::VAArg:
+			{
+				return true;
+			}
+			case Instruction::Call:
+			{
+				const CallBase& CB = cast<CallBase>(I);
+				if (const Function* FF = dyn_cast<Function>(CB.getCalledFunction()))
+				{
+					//Recursive calls might be handled, but require the mapping to be tied to the ExecutionContext
+					if (isAlreadyInCallStack(FF))
+						break;
+
+					if (FF->isDeclaration())
+						break;
+
+					uint32_t argIndex=0;
+					for (auto& op : CB.args())
+					{
+						if (argIndex >= FF->getFunctionType()->getNumParams())
+						{
+							//It means it's a VAArg, break from the inner loop
+							break;
+						}
+						assignToMapsIfComputed(FF->getArg(argIndex), op);
+						argIndex++;
+					}
+					return false;
+				}
+				break;
+			}
+			case Instruction::Load:
+			{
+				const LoadInst& LI = cast<LoadInst>(I);
+
+				GenericValue SRC = getOperandValue(LI.getPointerOperand());
+				GenericValue *Ptr = (GenericValue*)GVTORP(SRC);
+
+				for (const auto& interval : validLoadIntervals)
+				{
+					if ((long long)Ptr >= interval.first && (long long)Ptr < interval.second)
+						return false;
+				}
+
+				// TODO(carlo): Possibly even Alloca might be proven valid
+				break;
+			}
+			default:
+			{
+				// TODO(carlo): review PtrToInt or IntToPtr
+				// TODO(carlo): actually the proper way would be have the default meaning it has to be skipped, so that the conservative answer is failure
+				return false;
+			}
+		}
+		return true;
+	}
+	void computeValidLoadIntervals(const Module& _module)
+	{
+		for (const auto& global : _module.globals())
+		{
+			const GlobalVariable* GV = dyn_cast<GlobalVariable>(&global);
+			if (!GV)
+				continue;
+			if (!GV->hasInitializer())
+				continue;
+			if (GV->isExternallyInitialized())
+				continue;
+			if (!GV->hasInternalLinkage())
+				continue;
+			if (!GV->hasName())
+				continue;
+			if (!GV->isConstant())
+				continue;
+
+			GenericValue SRC = getOperandValue(GV);
+			GenericValue *Ptr = (GenericValue*)GVTORP(SRC);
+			
+			validLoadIntervals.push_back({(long long)Ptr, ((long long)Ptr)+getDataLayout().getTypeAllocSize(GV->getInitializer()->getType())});
+		}
+
+		std::sort(validLoadIntervals.begin(), validLoadIntervals.end());
+	}
+	template <PartialInterpreter::VisitingPolicy policy>
+	void visitOuter(const llvm::Instruction& I) 
+	{
+		if (policy == REMOVE_VALUES)
+		{
+			removeFromMaps(&I);
+			return;
+		}
+		
+		if (const PHINode* phi = dyn_cast<PHINode>(&I))
+		{
+			if (const llvm::BasicBlock* from = getIncomingBB())
+			{
+				const llvm::Value* incomingVal = phi->getIncomingValueForBlock(from);
+				assert(incomingVal);
+				if (isValueComputed(incomingVal))
+				{
+					incomingsVal.push_back({phi, incomingVal});
+					return;
+				}
+			}
+			notComputedPhis.push_back(phi);
+			return;	
+		}
+		else
+		{
+			//While visiting the first NonPhi instruction we set the maps state correctly
```

Comment:
Please make it explicit that this is required to deal with the simultaneity of PHIs. By the way, it seems weird we need to special case this. The exact same problem applies to the normal interpreter (i.e. PHIs needs to be dealt with all together, and not one-by-one). May there be a better entry point for PHIs?